### PR TITLE
Support deepseek v1 and v2

### PR DESCRIPTION
**Pull request by kkimmk**
Supports DeepSeek models that run on a single GPU (deepseek-moe-16b-chat, DeepSeek-V2-Lite-Chat).

- [feat: introduce patch module for remote transformers and add patched for deepseek models](https://github.com/SqueezeBits/Torch-TRTLLM/commit/4f047f800ffe72d5982e89adb34a615aec9f808f)
  
- [fix: cast router logits to FP32 if required](https://github.com/SqueezeBits/Torch-TRTLLM/commit/631cdb69a8dd39d8ebcdaf0bc6e4b490a804e1be)
  
- [feat: improve shared expert handling in MoE to handle alternating MoE layers](https://github.com/SqueezeBits/Torch-TRTLLM/commit/dda1ebcd18188f0ac71fd185bb2809ddc1cbf5c5)
  - Exclude shared expert linears for intermediate size identification
  - Add marking for linears in shared expert layers
  - Add utility methods in TrailingReformatPath to traverse graph nodes
  - FOC-939: Now converts deepseek-moe-16b-chat properly
  
- [fix: add moe config properly for config file generation](https://github.com/SqueezeBits/Torch-TRTLLM/commit/aa4d09966efd99039285a9bd6af85b3c7ccbf45f)
  
- [refactor: introduce a rope node to represent the wrapped rope](https://github.com/SqueezeBits/Torch-TRTLLM/commit/3536b24be0e5c474f4e217a2e1f8dccbb111444e)
  
- [feat: make find_nearest available with both Subgraph and NodeSpecialization and add continue condition](https://github.com/SqueezeBits/Torch-TRTLLM/commit/3b9314501798a90c67686229b6dbe1adf5886e51)
  
- [fix: use the original constant folding method from dynamo](https://github.com/SqueezeBits/Torch-TRTLLM/commit/30144d3a437a6668848cb0fd7c0ae10408bcec0b)
  - the custom constant folding is not properly working with ops using
    getitems for its outputs, such as SplitWithSizes.
  
- [add: support split_with_sizes aten op](https://github.com/SqueezeBits/Torch-TRTLLM/commit/ad97dd942f89747813cc15cfc5f57e1b85e87c05)
  
- [feat: support Multi-Head Latent Attention and DeepseekV2(FOC-953, FOC-947)](https://github.com/SqueezeBits/Torch-TRTLLM/commit/e7fb94f1f3cd7520e34a51b712273ae36c621c3b)
  - Introduce MLA class.
  - Refactor `ReplaceSDPAByGPTAttentionPlugin` to handle both MHA and MLA.
  - Allow scale value different from the `default_scale`.
  - Support RoPE for deepseek-v2.
  - Move RoPE calculation to after obtaining attention configuration.
  - Add rules for setting MoE normalization mode.
  - Add a model patch to prevent decomposition failure.
  
- [chore: add debug methods for MLA weights debugging](https://github.com/SqueezeBits/Torch-TRTLLM/commit/5ae7e3aa4512a7f7eb55222b51623b3fdcd360e1)
  
- [chore: add deepseek lite models to the test script](https://github.com/SqueezeBits/Torch-TRTLLM/commit/1799da099c77cbb900035d3062d66a41887a1fb2)
  
- [Merge branch 'dev' into support-deepseek-v1-v2](https://github.com/SqueezeBits/Torch-TRTLLM/commit/6c189bf4dea6d33c24eb205fb3dd45820b86e3fe)
  
- [chore: set default cleanup option to false in the test script](https://github.com/SqueezeBits/Torch-TRTLLM/commit/6b44a5912f241606d92f4e0cb7af7aefe4c303e0)
  
- [fix: get intermediate size properly for qwen-moe](https://github.com/SqueezeBits/Torch-TRTLLM/commit/e476ba3b8e38f90d4ce348f6666a0157f96d5c2b)

### CODE REVIEW

File: src/ditto/types.py [hunk 1]
```diff
@@ -36,13 +36,13 @@
...
 SymbolicInteger = torch.SymInt | int
-SymbolicShape = tuple[SymbolicInteger, ...]  # type: ignore[valid-type]
+SymbolicShape = tuple[SymbolicInteger, ...]
 ShapeArg = list[int | Node]
 
 T = TypeVar("T")
 
 
-def verify(value: Any, *, as_type: type[T], coerce: bool = False, **config: Unpack[ConfigDict]) -> T | None:
+def verify(value: Any, *, as_type: Any, coerce: bool = False, **config: Unpack[ConfigDict]) -> Any:
```
Comment (jiwoong-choi) [inline src/ditto/types.py:45]: Let's keep it as it was and ignore the mypy's complaints

**Review by jiwoong-choi — CHANGES_REQUESTED** [review]

File: src/ditto/fx/passes/replace_sdpa_by_gpt_attention_plugin.py [hunk 2]
```diff
@@ -50,69 +65,106 @@ def call(self, graph_module: GraphModule) -> PassResult:
...
                 continue
 
-            qkv_proj, mha = qkv_proj_and_mha
-            layer_idx += 1
-            logger.debug(f"Replacing SDPA at layer {layer_idx} with {mha = }")
+            is_mla_enabled = isinstance(attn, MLA)
+            if not sdpa.is_eligible_for_gpt_attention_plugin(is_mla_enabled):
+                continue
+            if is_mla_enabled:
+                assert isinstance(attn, MLA)
```
[the remark this answers is not in the document]
Comment (kkimmk) [reply src/ditto/fx/passes/replace_sdpa_by_gpt_attention_plugin.py:77]: mypy is poor at detecting type narrowing :( `is_mla_enabled` is being used several times later also, but I'll just modify using `isinstance()`, as the suggestion.

File: src/ditto/fx/utils.py
Comment (junstar92) [inline src/ditto/fx/utils.py:1]: I used `max_depth` for BFS. So, it could be failed when there is no found node until reaching the `max_depth` level of BFS. But, in your case, it checks the depth of the individual nodes. It looks like it can show warning messages even though the target node is found.

**Review by junstar92 — APPROVED** [review]

File: src/ditto/fx/utils.py [hunk 3]
```diff
@@ -290,44 +295,95 @@ def find_output_node(graph_or_graph_module: Graph | GraphModule) -> Node:
...
+        specialize_func = node_type.configure_from
+    else:
+        raise TypeError(f"Invalid specialization type: {node_type}")
+
+    queue = [(from_node, 0)]
+    while queue:
+        node, depth = queue.pop(0)
+        if target_node := specialize_func(node):
+            return target_node  # type: ignore[return-value]
+        if break_if is not None and break_if(node):
```
Comment (huijjj) [inline src/ditto/fx/utils.py:376]:
```suggestion
    from .nodes import NodeSpecialization
    from .subgraphs import Subgraph

    if issubclass(node_type, NodeSpecialization):
        specialize_func = node_type.specialize_from
    elif issubclass(node_type, Subgraph):
        specialize_func = node_type.configure_from
    else:
        raise TypeError(f"Invalid specialization type: {node_type}")

    stack = [(from_node, 0)]
    while stack:
        node, d = stack.pop(-1)
        if target_node := specialize_func(node):
            return target_node
        if break_if(node):
            break
        if continue_if(node):
            continue
        if depth > max_depth:
            continue
        next = list(node.all_input_nodes if follow_parent else node.users)
        if follow_first_only:
            next = next[:1]
        stack.extend(zip(next, [d + 1] * len(next)))
    logger.warning(f"Failed to find nearest {node_type} from {from_node} within max depth {max_depth}")
    return None
```

**Review by jiwoong-choi — APPROVED** [review]

File: src/ditto/fx/nodes/rope.py
[the remark this answers is not in the document]
Comment (kkimmk) [reply src/ditto/fx/nodes/rope.py:1]: This node is for easier specialization of the wrapped rope nodes. And it can be expanded to include rope_config, which is currently parsed through metadata. I'll refactor those rope configs in a separate PR.

File: src/ditto/fx/utils.py
Comment (kkimmk) [reply src/ditto/fx/utils.py:1]: That's true. I just removed the warning, because it seems more appropriate to handle the case from the caller side.

**Comment by huijjj** [comment]
LGTM
